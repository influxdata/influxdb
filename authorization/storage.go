package authorization

import (
	"context"
	goerrors "errors"
	"fmt"
	"slices"

	"github.com/go-crypt/crypt"
	"github.com/influxdata/influxdb/v2"
	"github.com/influxdata/influxdb/v2/kit/platform"
	"github.com/influxdata/influxdb/v2/kit/platform/errors"
	"github.com/influxdata/influxdb/v2/kit/tracing"
	"github.com/influxdata/influxdb/v2/kv"
	influxdb2_algo "github.com/influxdata/influxdb/v2/pkg/crypt/algorithm/influxdb2"
	"github.com/influxdata/influxdb/v2/snowflake"
)

/*---
Token storage and verification

Storage of hashed tokens has been added as an optional feature. This stores only the hash of a token
in BoltDB. Token hashing is enabled with the `--use-hashed-tokens` option.

Upgrading the BoltDB schema is automatic on startup when using a new version of InfluxDB with token hashing support.
Additionally, raw tokens are automatically migrated to hashed tokens if `--use-hashed-tokens` is configured.
Due to the schema changes, to use a version of InfluxDB without hashed token support, a manual downgrade using
`influxd downgrade` must be run. Any tokens stored as hashed tokens will be unusable by the old version of InfluxDB
and must be reset or recreated.

The implementation has the following behaviors under different scenarios:
* Token hashing is enabled.
  * On startup and upgrade, any raw tokens in BoltDB are automatically hashed. The hashed token is
    stored and the raw token is removed. No raw tokens remain in BoltDB.
  * On downgrade, hashed tokens are deleted from BoltDB along with their indices. User confirmation
    is required to proceed if hashed tokens are present.
  * Token verification is performed by hashing the token and finding the hashed token in BoltDB.
  * New tokens generated are stored only as hashes.
  * When creating a backup, only hashed tokens are exported.
  * When restoring a backup, any raw tokens are converted to hashed tokens and stored only as hashes.
    No raw tokens are imported into BoltDB.
  * When listing tokens, the hash algorithm used (e.g. `SHA-512`) is returned instead of the hashed token value.
* Token hashing is disabled.
  * On upgrade and startup, no user visible action is taken. Any hashed tokens in BoltDB remain unchanged.
    The BoltDB store is updated to support hashed tokens, but no existing tokens are migrated.
  * On downgrade, hashed tokens are deleted from BoltDB along with their indices. User confirmation
    is required to proceed if hashed tokens are present.
  * Token verification is performed by looking up the raw token value, if the provided token is not in PHC format.
    If the raw token value is not found, then the hashed token value is calculated and token lookup attempted again.
  * New tokens generated are stored as raw tokens.
  * When creating a backup, tokens are exported in the format found in BoltDB (raw or hashed).
  * When restoring a backup, both raw tokens and hashed tokens are restored unchanged.
  * When listing tokens, raw tokens in BoltDB are returned. Hashed tokens in BoltDB are returned as
    the hash algorithm (e.g. `SHA-512``) instead of the hashed token value.
* Downgrading to an older InfluxDB without hashed token support with a BoltDB containing hashed tokens.
  * Downgrading requires a manual `influxd downgrade` command. If hashed tokens are found in the
    BoltDB, user confirmation is required. The user can also list impacted tokens. When downgrade is
	complete, all hashed tokens have been deleted from BoltDB along with their indices. The tokens with
	deleted hashes are no longer useable.
  * Operations are as usual in InfluxDB 2.7.

The hashed tokens in `Authorization.HashedToken` are stored in PHC format. PHC allows specifying the
algorithm used and any parameters. This allows gradual token algorithm transitions if new token hashing
algorithms are added in the future. PHC is used over MCF (Modular Crypt Format) because PHC is more
flexible and MCF does not support our chosen hash scheme.

When token hashing is enabled, on every startup (not just upgrade) InfluxDB scans the BoltDB for raw tokens in
the `Authorization.Token` field. When found, a hash is immediately stored in the `Authorization.HashedToken`
field and the `Authorization.Token` field is cleared. The token index is also updated to use the hashed
token value instead of the raw token value. This migration must occur on every startup and not just upgrades
because hashed tokens can be turned on and off with configuration.

When a backup is made, the format stored in BoltDB is exported as-is. If hashed tokens are enabled, only hashed
tokens are exported since only hashed tokens are stored. Without enabling hashed tokens, a mix of raw and
hashed tokens may be present in the backup.

When token hashing is enabled and a backup is restored, raw tokens are hashed before importing
into BoltDB. Raw tokens are not stored.

To verify tokens when hashed tokens are enabled, the presented token's hash is calculated and used
for token index lookup. The rest of the authorization flow is unchanged.

To verify tokens when hashed tokens are disabled, the an attempt is made to parse the presented token as
PHC. If the parse succeeds, the access is denied. This prevents an attack described below. After this check,
the presented raw token is used to lookup the token in the raw token index. If found, authorization proceeds
as normal. Otherwise, the token hash is calculated and used to lookup the token in the hashed token index.
A second check is then done on the authorization record token or token hash matches the presented token.
If found, authorization proceeds as normal.

The hashed token index is separate from the raw token index. Newer versions also verify that the token
is not a valid PHC string before starting authorization. This prevents the following attack:
1. Hashed token is extracted from BoltDB.
2. Token hashing is disabled.
3. The hashed token is presented to the API, which will misinterpret it as a raw token and allow access.

The token hashing algorithm is SHA-512. This provides a good level of security and is allowed by FIPS 140-2.
Because the token hashes must be useable as index lookups, salted password hashes (e.g. bcrypt, PBKDF2, Argon)
can not be used.

A potential future security would be optionally storing "peppered" hashes. This would require retrieving
the pepper key from outside of BoltDB, for example from Vault.

When listing tokens, hashed tokens are listed as "REDACTED" instead of the hashed
token value. Raw token values are returned as in previous versions.

---*/

const MaxIDGenerationN = 100
const ReservedIDs = 1000

var (
	ErrReadOnly = goerrors.New("authorization store is read-only")
)

var (
	authBucket      = []byte("authorizationsv1")
	authIndex       = []byte("authorizationindexv1")
	hashedAuthIndex = []byte("authorizationhashedindexv1")
)

type Store struct {
	kvStore kv.Store
	IDGen   platform.IDGenerator
	hasher  *AuthorizationHasher

	// Indicates if tokens should be stored in hashed PHC format.
	useHashedTokens bool

	// Indicates if Store is read-only.
	readOnly bool

	// ignoreMissingHashIndex indicates if missing hash indices in store should be ignored.
	// This is almost exclusively for testing.
	ignoreMissingHashIndex bool
}

type storePlusOptions struct {
	*Store
	hasherVariantName string
}

type StoreOption func(*storePlusOptions)

func WithAuthorizationHasher(hasher *AuthorizationHasher) StoreOption {
	return func(s *storePlusOptions) {
		s.hasher = hasher
	}
}

func WithAuthorizationHashVariantName(name string) StoreOption {
	return func(s *storePlusOptions) {
		s.hasherVariantName = name
	}
}

func WithReadOnly(readOnly bool) StoreOption {
	return func(s *storePlusOptions) {
		s.readOnly = readOnly
	}
}

func WithIgnoreMissingHashIndex(allowMissing bool) StoreOption {
	return func(s *storePlusOptions) {
		s.ignoreMissingHashIndex = allowMissing
	}
}

// NewStore creates a new authorization.Store object. kvStore is the underlying key-value store.
func NewStore(ctx context.Context, kvStore kv.Store, useHashedTokens bool, opts ...StoreOption) (*Store, error) {
	s := &storePlusOptions{
		Store: &Store{
			kvStore:         kvStore,
			IDGen:           snowflake.NewDefaultIDGenerator(),
			useHashedTokens: useHashedTokens,
		},
		hasherVariantName: DefaultHashVariantName,
	}

	for _, o := range opts {
		o(s)
	}

	if err := s.setup(ctx); err != nil {
		return nil, err
	}

	if s.hasher == nil {
		hasher, err := s.autogenerateHasher(ctx, s.hasherVariantName)
		if err != nil {
			return nil, err
		}
		s.hasher = hasher
	}

	// Perform hashed token migration if needed. This can not be performed by the migration service
	// because it requires configuration, and the migration service is more concerned with schema
	// and does not have configuration.
	if err := s.hashedTokenMigration(ctx); err != nil {
		return nil, fmt.Errorf("error during hashed token migration: %w", err)
	}

	return s.Store, nil
}

// autogenerateHasher generates an AuthorizationHasher that hashes using variantName.
// The decoders include variantName plus any other variants that are detected in the
// store.
func (s *Store) autogenerateHasher(ctx context.Context, variantName string) (*AuthorizationHasher, error) {
	// Determine which variants are present in the store.
	tempDecoder := crypt.NewDecoder()
	if err := influxdb2_algo.RegisterDecoder(tempDecoder); err != nil {
		return nil, fmt.Errorf("error registering test decoders for authorization store: %w", err)
	}

	var auths []*influxdb.Authorization
	err := s.View(ctx, func(tx kv.Tx) error {
		as, err := s.ListAuthorizations(ctx, tx, influxdb.AuthorizationFilter{})
		if err != nil {
			return err
		}
		auths = as
		return nil
	})
	if err != nil {
		return nil, fmt.Errorf("error fetching authorization records for hash variant inventory: %w", err)
	}

	foundVariants := make(map[influxdb2_algo.Variant]struct{})
	for _, a := range auths {
		if a.HashedToken != "" {
			digest, err := tempDecoder.Decode(a.HashedToken)
			if err == nil {
				if influxdbDigest, ok := digest.(*influxdb2_algo.Digest); ok {
					foundVariants[influxdbDigest.Variant] = struct{}{}
				}
			}
		}
	}

	var decoderVariants []influxdb2_algo.Variant
	// Make sure we have the hasher variant we will make in there and that it is first in the list.
	hasherVariant := influxdb2_algo.NewVariant(variantName)
	decoderVariants = append(decoderVariants, hasherVariant)
	delete(foundVariants, hasherVariant)
	for variant := range foundVariants {
		decoderVariants = append(decoderVariants, variant)
	}

	hasher, err := NewAuthorizationHasher(WithHasherVariant(hasherVariant), WithDecoderVariants(decoderVariants))
	if err != nil {
		return nil, fmt.Errorf("error creating authorization hasher for authorization store: %w", err)
	}

	return hasher, nil
}

// hashedTokenMigration migrates any unhashed tokens in the store to hashed tokens.
func (s *Store) hashedTokenMigration(ctx context.Context) error {
	if !s.useHashedTokens || s.readOnly {
		return nil
	}

	// Figure out which authorization records need to be updated.
	var authsNeedingUpdate []*influxdb.Authorization
	err := s.View(ctx, func(tx kv.Tx) error {
		s.forEachAuthorization(ctx, tx, nil, func(a *influxdb.Authorization) bool {
			if a.HashedToken == "" && a.Token != "" {
				authsNeedingUpdate = append(authsNeedingUpdate, a)
			}
			return true
		})
		return nil
	})
	if err != nil {
		return err
	}

	for batch := range slices.Chunk(authsNeedingUpdate, 100) {
		err := s.Update(ctx, func(tx kv.Tx) error {
			// Now update them. This really seems too simple, but s.UpdateAuthorization() is magical.
			for _, a := range batch {
				if _, err := s.UpdateAuthorization(ctx, tx, a.ID, a); err != nil {
					return err
				}
			}
			return nil
		})
		if err != nil {
			return fmt.Errorf("error migrating hashed tokens: %w", err)
		}
	}

	return nil
}

// View opens up a transaction that will not write to any data. Implementing interfaces
// should take care to ensure that all view transactions do not mutate any data.
func (s *Store) View(ctx context.Context, fn func(kv.Tx) error) error {
	return s.kvStore.View(ctx, fn)
}

// Update opens up a transaction that will mutate data.
func (s *Store) Update(ctx context.Context, fn func(kv.Tx) error) error {
	if s.readOnly {
		return ErrReadOnly
	}
	return s.kvStore.Update(ctx, fn)
}

func (s *Store) setup(ctx context.Context) error {
	return s.View(ctx, func(tx kv.Tx) error {
		if _, err := tx.Bucket(authBucket); err != nil {
			return err
		}
		if _, err := authIndexBucket(tx); err != nil {
			return err
		}
		if _, err := hashedAuthIndexBucket(tx); err != nil {
			if goerrors.Is(err, kv.ErrBucketNotFound) {
				if !s.ignoreMissingHashIndex || (s.useHashedTokens && !s.readOnly) {
					return fmt.Errorf("missing required index, upgrade required: %w", err)
				}
			} else {
				return err
			}
		}

		return nil
	})
}

// generateSafeID attempts to create ids for buckets
// and orgs that are without backslash, commas, and spaces, BUT ALSO do not already exist.
func (s *Store) generateSafeID(ctx context.Context, tx kv.Tx, bucket []byte) (platform.ID, error) {
	for i := 0; i < MaxIDGenerationN; i++ {
		id := s.IDGen.ID()

		// TODO: this is probably unnecessary but for testing we need to keep it in.
		// After KV is cleaned out we can update the tests and remove this.
		if id < ReservedIDs {
			continue
		}

		err := s.uniqueID(ctx, tx, bucket, id)
		if err == nil {
			return id, nil
		}

		if err == NotUniqueIDError {
			continue
		}

		return platform.InvalidID(), err
	}
	return platform.InvalidID(), ErrFailureGeneratingID
}

func (s *Store) uniqueID(ctx context.Context, tx kv.Tx, bucket []byte, id platform.ID) error {
	span, _ := tracing.StartSpanFromContext(ctx)
	defer span.Finish()

	encodedID, err := id.Encode()
	if err != nil {
		return &errors.Error{
			Code: errors.EInvalid,
			Err:  err,
		}
	}

	b, err := tx.Bucket(bucket)
	if err != nil {
		return err
	}

	_, err = b.Get(encodedID)
	if kv.IsNotFound(err) {
		return nil
	}

	return NotUniqueIDError
}
