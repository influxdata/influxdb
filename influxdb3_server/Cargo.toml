[package]
name = "influxdb3_server"
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true

[dependencies]
# Core Crates
authz.workspace = true
data_types.workspace = true
datafusion_util.workspace = true
influxdb-line-protocol.workspace = true
iox_catalog.workspace = true
iox_http.workspace = true
iox_query.workspace = true
iox_query_params.workspace = true
iox_query_influxql.workspace = true
iox_system_tables.workspace = true
iox_time.workspace = true
metric.workspace = true
metric_exporters.workspace = true
observability_deps.workspace = true
schema.workspace = true
service_common.workspace = true
service_grpc_flight.workspace = true
trace.workspace = true
trace_exporters.workspace = true
trace_http.workspace = true
tracker.workspace = true

# Local Deps
influxdb3_catalog = { path = "../influxdb3_catalog" }
influxdb3_write = { path = "../influxdb3_write" }
influxdb3_process = { path = "../influxdb3_process", default-features = false }
influxdb3_wal = { path = "../influxdb3_wal"}
iox_query_influxql_rewrite = { path = "../iox_query_influxql_rewrite" }

# crates.io Dependencies
anyhow.workspace = true
arrow.workspace = true
arrow-array.workspace = true
arrow-csv.workspace = true
arrow-flight.workspace = true
arrow-json.workspace = true
arrow-schema.workspace = true
async-trait.workspace = true
base64.workspace = true
bytes.workspace = true
chrono.workspace = true
csv.workspace = true
datafusion.workspace = true
flate2.workspace = true
futures.workspace = true
hex.workspace = true
hyper.workspace = true
mime.workspace = true
object_store.workspace = true
parking_lot.workspace = true
pin-project-lite.workspace = true
secrecy.workspace = true
serde.workspace = true
serde_json.workspace = true
serde_urlencoded.workspace = true
sha2.workspace = true
thiserror.workspace = true
tokio.workspace = true
tokio-util.workspace = true
tonic.workspace = true
tower.workspace = true
unicode-segmentation.workspace = true

[dev-dependencies]
# Core Crates
parquet.workspace = true
parquet_file.workspace = true
test_helpers.workspace = true

# crates.io crates
http.workspace = true
hyper.workspace = true
test-log.workspace = true
urlencoding.workspace = true
pretty_assertions.workspace = true
