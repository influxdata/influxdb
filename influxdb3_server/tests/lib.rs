use std::{collections::HashMap, num::NonZeroUsize, sync::Arc, time::Duration};

use arrow_array::RecordBatch;
use data_types::NamespaceName;
use datafusion::assert_batches_sorted_eq;
use futures::TryStreamExt;
use influxdb3_cache::{distinct_cache::DistinctCacheProvider, last_cache::LastCacheProvider};
use influxdb3_catalog::catalog::Catalog;
use influxdb3_internal_api::query_executor::QueryExecutor;
use influxdb3_server::query_executor::{CreateQueryExecutorArgs, QueryExecutorImpl};
use influxdb3_shutdown::ShutdownManager;
use influxdb3_sys_events::SysEventStore;
use influxdb3_telemetry::store::TelemetryStore;
use influxdb3_wal::{Gen1Duration, WalConfig};
use influxdb3_write::{
    Precision, WriteBuffer,
    persister::Persister,
    write_buffer::{N_SNAPSHOTS_TO_LOAD_ON_START, WriteBufferImpl, WriteBufferImplArgs},
};
use iox_query::exec::{DedicatedExecutor, Executor, ExecutorConfig, PerQueryMemoryPoolConfig};
use iox_time::{MockProvider, Time, TimeProvider};
use metric::Registry;
use object_store::{ObjectStore, memory::InMemory};
use parquet_file::storage::{ParquetStorage, StorageId};

/// Test that when duplicates are written into the write buffer, the query results do not contain
/// duplicate rows
#[test_log::test(tokio::test)]
async fn test_deduplicate_rows_in_write_buffer_memory() {
    let wal_config = WalConfig {
        gen1_duration: Gen1Duration::new_1m(),
        max_write_buffer_size: 100,
        flush_interval: Duration::from_millis(10),
        // use a large snapshot size so that queries are served only by in-memory buffer chunks:
        snapshot_size: 100,
        ..Default::default()
    };
    let service = TestService::setup(wal_config).await;

    // do the same writes every time:
    let writes = [
        TestWrite::lp("bar val=1", 1),
        TestWrite::lp("bar val=2", 2),
        TestWrite::lp("bar val=3", 3),
    ];

    // query results should always be the same:
    let expected = [
        "+---------------------+-----+",
        "| time                | val |",
        "+---------------------+-----+",
        "| 1970-01-01T00:00:01 | 1.0 |",
        "| 1970-01-01T00:00:02 | 2.0 |",
        "| 1970-01-01T00:00:03 | 3.0 |",
        "+---------------------+-----+",
    ];

    // do the writes several times:
    for _ in 0..5 {
        service.do_writes("foo", writes.clone()).await;
        let batches = service.query_sql("foo", "select * from bar").await;
        assert_batches_sorted_eq!(expected, &batches);
    }

    // check the explain plan:
    let batches = service.query_sql("foo", "explain select * from bar").await;
    let plan = arrow::util::pretty::pretty_format_batches(&batches)
        .unwrap()
        .to_string();
    // checking that the DeduplicateExec is above the RecordBatchesExec in the query plan:
    insta::assert_snapshot!(plan);
}

/// Test that when duplicates are written into the write buffer, across snapshot points, i.e., so
/// that there are duplicate rows contained across multiple parquet files, the query results do not
/// contain duplicate rows
#[test_log::test(tokio::test)]
async fn test_deduplicate_rows_in_write_buffer_parquet() {
    let wal_config = WalConfig {
        gen1_duration: Gen1Duration::new_1m(),
        max_write_buffer_size: 100,
        flush_interval: Duration::from_millis(10),
        // uses a small snapshot size to get parquet persisted early, and therefore have queries
        // serve chunks from both in-memory buffer and persisted parquet
        snapshot_size: 1,
        ..Default::default()
    };
    let service = TestService::setup(wal_config).await;

    // do the same 3 writes every time:
    let writes = [
        TestWrite::lp("bar val=1", 1),
        TestWrite::lp("bar val=2", 2),
        TestWrite::lp("bar val=3", 3),
    ];

    // query result should always be the same:
    let expected = [
        "+---------------------+-----+",
        "| time                | val |",
        "+---------------------+-----+",
        "| 1970-01-01T00:00:01 | 1.0 |",
        "| 1970-01-01T00:00:02 | 2.0 |",
        "| 1970-01-01T00:00:03 | 3.0 |",
        "+---------------------+-----+",
    ];

    // do the writes and query several times:
    for i in 1..5 {
        // write then wait for a snapshot to ensure parquet
        // is persisted:
        service.do_writes("foo", writes.clone()).await;
        service.wait_for_snapshot_sequence(i).await;

        let batches = service.query_sql("foo", "select * from bar").await;
        assert_batches_sorted_eq!(expected, &batches);
    }

    // check the query plan to ensure that deduplication is taking place:
    let batches = service.query_sql("foo", "explain select * from bar").await;
    let plan = arrow::util::pretty::pretty_format_batches(&batches)
        .unwrap()
        .to_string();
    // looking to see that the `DeduplicateExec` is above the `ParquetExec` to ensure data from
    // multiple parquet files is deduplicated:
    insta::assert_snapshot!(plan);
}

#[test_log::test(tokio::test)]
async fn test_deduplicate_rows_in_write_buffer_both_memory_and_parquet() {
    let wal_config = WalConfig {
        gen1_duration: Gen1Duration::new_1m(),
        max_write_buffer_size: 100,
        flush_interval: Duration::from_millis(10),
        // uses a small snapshot size to get parquet persisted early, and therefore have queries
        // serve chunks from both in-memory buffer and persisted parquet
        snapshot_size: 1,
        ..Default::default()
    };
    let service = TestService::setup(wal_config).await;

    service
        .do_writes(
            "foo",
            [
                TestWrite::lp("bar value=false", 1),
                TestWrite::lp("bar value=false", 2),
                TestWrite::lp("bar value=false", 3),
            ],
        )
        .await;

    service.wait_for_snapshot_sequence(1).await;

    let batches = service.query_sql("foo", "select * from bar").await;
    assert_batches_sorted_eq!(
        [
            "+---------------------+-------+",
            "| time                | value |",
            "+---------------------+-------+",
            "| 1970-01-01T00:00:01 | false |", // note that this is `false`
            "| 1970-01-01T00:00:02 | false |",
            "| 1970-01-01T00:00:03 | false |",
            "+---------------------+-------+",
        ],
        &batches
    );

    // write a duplicate row, but don't trigger a snapshot, so subsequent query will be served
    // by chunks in both memory buffer and persisted parquet; this also flips the `value` to `true`
    // to show that the most recent written line appears in the query result:
    service
        .do_writes("foo", [TestWrite::lp("bar value=true", 1)])
        .await;

    let batches = service.query_sql("foo", "select * from bar").await;
    assert_batches_sorted_eq!(
        [
            "+---------------------+-------+",
            "| time                | value |",
            "+---------------------+-------+",
            "| 1970-01-01T00:00:01 | true  |", // note that this is now `true`
            "| 1970-01-01T00:00:02 | false |",
            "| 1970-01-01T00:00:03 | false |",
            "+---------------------+-------+",
        ],
        &batches
    );

    let batches = service.query_sql("foo", "explain select * from bar").await;
    // There should be a union of a ParquetExec and RecordBatchesExec that feeds into a
    // DeduplicateExec in the query plan, i.e., there is both in-memory buffer chunks and persisted
    // parquet chunks:
    let plan = arrow::util::pretty::pretty_format_batches(&batches)
        .unwrap()
        .to_string();
    insta::assert_snapshot!(plan);
}

struct TestService {
    query_executor: Arc<dyn QueryExecutor>,
    write_buffer: Arc<dyn WriteBuffer>,
    _time_provider: Arc<dyn TimeProvider>,
    _metrics: Arc<Registry>,
    _object_store: Arc<dyn ObjectStore>,
}

impl TestService {
    async fn setup(wal_config: WalConfig) -> Self {
        let node_id = "test-node";
        let object_store = Arc::new(InMemory::new());
        let time_provider = Arc::new(MockProvider::new(Time::from_timestamp_nanos(0)));
        let metrics = Arc::new(Registry::new());
        let catalog = Arc::new(
            Catalog::new(
                node_id,
                Arc::clone(&object_store) as _,
                Arc::clone(&time_provider) as _,
                Arc::clone(&metrics),
            )
            .await
            .unwrap(),
        );
        let parquet_store =
            ParquetStorage::new(Arc::clone(&object_store) as _, StorageId::from("influxdb3"));
        let exec = Arc::new(Executor::new_with_config_and_executor(
            ExecutorConfig {
                target_query_partitions: NonZeroUsize::new(1).unwrap(),
                object_stores: [&parquet_store]
                    .into_iter()
                    .map(|store| (store.id(), Arc::clone(store.object_store())))
                    .collect(),
                metric_registry: Arc::clone(&metrics),
                mem_pool_size: usize::MAX,
                per_query_mem_pool_config: PerQueryMemoryPoolConfig::Disabled,
                heap_memory_limit: None,
            },
            DedicatedExecutor::new_testing(),
        ));
        let persister = Arc::new(Persister::new(
            Arc::clone(&object_store) as _,
            node_id,
            Arc::clone(&time_provider) as _,
        ));
        let write_buffer: Arc<dyn WriteBuffer> = WriteBufferImpl::new(WriteBufferImplArgs {
            persister: Arc::clone(&persister),
            catalog: Arc::clone(&catalog),
            last_cache: LastCacheProvider::new_from_catalog(Arc::clone(&catalog))
                .await
                .unwrap(),
            distinct_cache: DistinctCacheProvider::new_from_catalog(
                Arc::clone(&time_provider) as _,
                Arc::clone(&catalog),
            )
            .await
            .unwrap(),
            time_provider: Arc::clone(&time_provider) as _,
            executor: Arc::clone(&exec),
            wal_config,
            parquet_cache: None,
            metric_registry: Arc::clone(&metrics),
            snapshotted_wal_files_to_keep: 100,
            query_file_limit: None,
            shutdown: ShutdownManager::new_testing().register(),
            n_snapshots_to_load_on_start: N_SNAPSHOTS_TO_LOAD_ON_START,
            wal_replay_concurrency_limit: Some(1),
        })
        .await
        .unwrap();
        let sys_events_store = Arc::new(SysEventStore::new(Arc::clone(&time_provider) as _));
        let telemetry_store =
            TelemetryStore::new_without_background_runners(None, Arc::clone(&catalog) as _);
        let mut datafusion_config = HashMap::new();
        datafusion_config.insert(
            "iox.hint_known_object_size_to_object_store".to_string(),
            false.to_string(),
        );
        let query_executor = Arc::new(QueryExecutorImpl::new(CreateQueryExecutorArgs {
            catalog: Arc::clone(&catalog),
            write_buffer: Arc::clone(&write_buffer) as _,
            exec: Arc::clone(&exec),
            metrics: Arc::clone(&metrics),
            datafusion_config: Arc::new(datafusion_config),
            query_log_size: 10,
            telemetry_store: Arc::clone(&telemetry_store),
            sys_events_store: Arc::clone(&sys_events_store),
            started_with_auth: false,
            time_provider: Arc::clone(&time_provider) as _,
        }));

        Self {
            query_executor,
            write_buffer,
            _time_provider: time_provider,
            _metrics: metrics,
            _object_store: object_store,
        }
    }
}

#[derive(Clone)]
struct TestWrite<LP> {
    lp: LP,
    time_seconds: i64,
}

impl<LP> TestWrite<LP>
where
    LP: AsRef<str>,
{
    fn lp(lp: LP, time_seconds: i64) -> Self {
        Self { lp, time_seconds }
    }
}

impl TestService {
    async fn do_writes<LP: AsRef<str>>(
        &self,
        db: &'static str,
        writes: impl IntoIterator<Item = TestWrite<LP>>,
    ) {
        for w in writes {
            self.write_buffer
                .write_lp(
                    NamespaceName::new(db).unwrap(),
                    w.lp.as_ref(),
                    Time::from_timestamp_nanos(w.time_seconds * 1_000_000_000),
                    false,
                    Precision::Nanosecond,
                    false,
                )
                .await
                .unwrap();
        }
    }

    async fn query_sql(&self, db: &str, query_str: &str) -> Vec<RecordBatch> {
        let stream = self
            .query_executor
            .query_sql(db, query_str, None, None, None)
            .await
            .unwrap();
        stream.try_collect().await.unwrap()
    }

    async fn wait_for_snapshot_sequence(&self, wait_for: u64) {
        let mut count = 0;
        loop {
            let last = self
                .write_buffer
                .wal()
                .last_snapshot_sequence_number()
                .await
                .as_u64();
            if last >= wait_for {
                break;
            }
            count += 1;
            tokio::time::sleep(Duration::from_millis(10)).await;
            if count > 100 {
                panic!("waited too long for a snapshot with sequence {wait_for}");
            }
        }
    }
}
