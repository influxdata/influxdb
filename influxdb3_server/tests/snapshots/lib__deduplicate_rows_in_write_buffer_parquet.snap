---
source: influxdb3_server/tests/lib.rs
expression: plan
---
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                                                                                                                                   |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | TableScan: bar projection=[time, val]                                                                                                                                                                                                                                                                                                                                                  |
| physical_plan | ProjectionExec: expr=[time@0 as time, val@1 as val]                                                                                                                                                                                                                                                                                                                                    |
|               |   DeduplicateExec: [time@0 ASC]                                                                                                                                                                                                                                                                                                                                                        |
|               |     SortExec: expr=[time@0 ASC, __chunk_order@2 ASC], preserve_partitioning=[false]                                                                                                                                                                                                                                                                                                    |
|               |       ParquetExec: file_groups={1 group: [[test-node/dbs/foo-1/bar-0/1970-01-01/00-00/0000000003.parquet, test-node/dbs/foo-1/bar-0/1970-01-01/00-00/0000000006.parquet, test-node/dbs/foo-1/bar-0/1970-01-01/00-00/0000000009.parquet, test-node/dbs/foo-1/bar-0/1970-01-01/00-00/0000000012.parquet]]}, projection=[time, val, __chunk_order], output_ordering=[__chunk_order@2 ASC] |
|               |                                                                                                                                                                                                                                                                                                                                                                                        |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
